### PR TITLE
[JBPM-9735] Improve jbpm engine Message and Signal

### PR DESCRIPTION
**Pull request by elguardian**
Jira: https://issues.redhat.com/browse/JBPM-9735

### CODE REVIEW

**Comment by elguardian** [comment]
Jenkins retest this please

**Comment by elguardian** [comment]
Jenkins retest this please

**Review by gmunozfe — CHANGES_REQUESTED** [review]
Looks good to me, huge and fabulous work @elguardian !
Check some typos (especially with metadata) and other minor comments.
It'd be great to add some more testing regarding edge conditions or events catching in subprocesses.

File: jbpm-flow/src/main/java/org/jbpm/process/core/correlation/Correlation.java [hunk 1]
```diff
@@ -0,0 +1,69 @@
...
+    private CorrelationProperties processCorrelationProperties;
+
+    public Correlation(String id, String name) {
+        this.id = id;
+        this.name = name;
+        this.correlationPropertiesByMessage = new HashMap<>();
+    }
+
+    public CorrelationProperties getMessageCorrelationFor(String messageRef) {
+        return correlationPropertiesByMessage.computeIfAbsent(messageRef, (ref) -> new CorrelationProperties());
```
[the remark this answers is not in the document]
Comment (elguardian) [reply jbpm-flow/src/main/java/org/jbpm/process/core/correlation/Correlation.java:40]: done

File: jbpm-flow/src/main/java/org/jbpm/process/core/correlation/CorrelationInstance.java [hunk 2]
```diff
@@ -0,0 +1,90 @@
...
+            if (other.name != null)
+                return false;
+        } else if (!name.equals(other.name))
+            return false;
+
+        if ((other.properties == null && properties != null) || (other.properties != null && properties == null)) {
+            return false;
+        }
+
+        if(properties.size() != other.properties.size()) {
```
[the remark this answers is not in the document]
Comment (elguardian) [reply jbpm-flow/src/main/java/org/jbpm/process/core/correlation/CorrelationInstance.java:73]: that cannot happen as constructor will have always an empty map at least

File: jbpm-flow/src/main/java/org/jbpm/process/core/correlation/Message.java [hunk 3]
```diff
@@ -0,0 +1,43 @@
...
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.jbpm.process.core.correlation;
+
+public class Message {
```
[the remark this answers is not in the document]
Comment (elguardian) [reply jbpm-flow/src/main/java/org/jbpm/process/core/correlation/Message.java:19]: yep

File: jbpm-bpmn2/src/main/java/org/jbpm/bpmn2/core/CorrelationKey.java [hunk 4]
```diff
@@ -0,0 +1,52 @@
...
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.jbpm.bpmn2.core;
+
+import java.util.ArrayList;
+import java.util.List;
+
+public class CorrelationKey {
```
[the remark this answers is not in the document]
Comment (elguardian) [reply jbpm-bpmn2/src/main/java/org/jbpm/bpmn2/core/CorrelationKey.java:22]: yep

File: jbpm-bpmn2/src/main/java/org/jbpm/bpmn2/core/Collaboration.java [hunk 5]
```diff
@@ -0,0 +1,66 @@
...
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.jbpm.bpmn2.core;
+
+import java.util.ArrayList;
+import java.util.List;
+
+public class Collaboration {
```
[the remark this answers is not in the document]
Comment (elguardian) [reply jbpm-bpmn2/src/main/java/org/jbpm/bpmn2/core/Collaboration.java:22]: yep

File: jbpm-bpmn2/src/main/java/org/jbpm/bpmn2/core/CorrelationSubscription.java [hunk 6]
```diff
@@ -0,0 +1,54 @@
...
+ * limitations under the License.
+ */
+
+
+package org.jbpm.bpmn2.core;
+
+import java.util.HashMap;
+import java.util.Map;
+
+public class CorrelationSubscription {
```
[the remark this answers is not in the document]
Comment (elguardian) [reply jbpm-bpmn2/src/main/java/org/jbpm/bpmn2/core/CorrelationSubscription.java:23]: yep

File: jbpm-bpmn2/src/main/java/org/jbpm/bpmn2/xml/CollaborationHandler.java [hunk 7]
```diff
@@ -0,0 +1,90 @@
...
+    @Override
+    public Object end(String uri, String localName, ExtensibleXmlParser parser) throws SAXException {
+        Element element = parser.endElementBuilder();
+        Collaboration collaboration = (Collaboration) parser.getCurrent();
+        buildCorrelationKeys(collaboration, element.getChildNodes());
+        return null;
+    }
+
+    private void buildCorrelationKeys(Collaboration collaboration , NodeList childNodes) {
+        for(int i = 0; i < childNodes.getLength(); i++) {
```
[the remark this answers is not in the document]
Comment (elguardian) [reply jbpm-bpmn2/src/main/java/org/jbpm/bpmn2/xml/CollaborationHandler.java:58]: done

File: jbpm-bpmn2/src/main/java/org/jbpm/bpmn2/xml/CollaborationHandler.java [hunk 8]
```diff
@@ -0,0 +1,90 @@
...
+            }
+        }
+    }
+
+    private List<String> buildPropertiesRef(NodeList childNodes) {
+        List<String> propertiesRef = new ArrayList<>();
+        for(int i = 0; i < childNodes.getLength(); i++) {
+            Node node = childNodes.item(i);
+            if("correlationPropertyRef".equals(node.getNodeName())) {
+                propertiesRef.add(node.getTextContent().trim());
```
[the remark this answers is not in the document]
Comment (elguardian) [reply jbpm-bpmn2/src/main/java/org/jbpm/bpmn2/xml/CollaborationHandler.java:79]: removed

File: jbpm-bpmn2/src/main/java/org/jbpm/bpmn2/xml/HandlerUtil.java [hunk 9]
```diff
@@ -0,0 +1,78 @@
...
+    public static Map<String, Message> messages(ExtensibleXmlParser parser) {
+        Map<String, Message> messages = (Map<String, Message>) ((ProcessBuildData) parser.getData()).getMetaData("Messages");
+        if(messages == null) {
+            messages = new HashMap<>();
+            ((ProcessBuildData) parser.getData()).setMetaData("Messages", messages);
+        }
+        return messages;
+    }
+
+    public static Map<String, ItemDefinition> defintions(ExtensibleXmlParser parser) {
```
[the remark this answers is not in the document]
Comment (elguardian) [reply jbpm-bpmn2/src/main/java/org/jbpm/bpmn2/xml/HandlerUtil.java:43]: done

File: jbpm-bpmn2/src/main/java/org/jbpm/bpmn2/xml/HandlerUtil.java [hunk 10]
```diff
@@ -0,0 +1,78 @@
+/*
+ * Copyright 2017 Red Hat, Inc. and/or its affiliates.
```
[the remark this answers is not in the document]
Comment (elguardian) [reply jbpm-bpmn2/src/main/java/org/jbpm/bpmn2/xml/HandlerUtil.java:2]: done

File: jbpm-bpmn2/src/main/java/org/jbpm/bpmn2/xml/HandlerUtil.java [hunk 11]
```diff
@@ -0,0 +1,78 @@
...
+        if(messages == null) {
+            messages = new HashMap<>();
+            ((ProcessBuildData) parser.getData()).setMetaData("Messages", messages);
+        }
+        return messages;
+    }
+
+    public static Map<String, ItemDefinition> defintions(ExtensibleXmlParser parser) {
+        Map<String, ItemDefinition> definitions = (Map<String, ItemDefinition>) ((ProcessBuildData) parser.getData()).getMetaData("ItemDefinitions");
+        if(definitions == null) {
```
[the remark this answers is not in the document]
Comment (elguardian) [reply jbpm-bpmn2/src/main/java/org/jbpm/bpmn2/xml/HandlerUtil.java:45]: done

File: jbpm-bpmn2/src/main/java/org/jbpm/bpmn2/xml/HandlerUtil.java [hunk 12]
```diff
@@ -0,0 +1,78 @@
...
+            definitions = new HashMap<>();
+            ((ProcessBuildData) parser.getData()).setMetaData("ItemDefinitions", definitions);
+        }
+        return definitions;
+    }
+
+    public static Map<String, CorrelationProperty> correlationProperties (ExtensibleXmlParser parser) {
+        Map<String, CorrelationProperty> properties = (Map<String, CorrelationProperty>) ((ProcessBuildData) parser.getData()).getMetaData("CorrelationProperties");
+        if(properties == null) {
```
[the remark this answers is not in the document]
Comment (elguardian) [reply jbpm-bpmn2/src/main/java/org/jbpm/bpmn2/xml/HandlerUtil.java:54]: done

File: jbpm-bpmn2/src/main/java/org/jbpm/bpmn2/xml/HandlerUtil.java [hunk 13]
```diff
@@ -0,0 +1,78 @@
...
+        if(collaborations == null) {
+            collaborations = new HashMap<>();
+            ((ProcessBuildData) parser.getData()).setMetaData("Collaborations", collaborations);
+        }
+        return collaborations;
+    }
+
+    public static Map<String, CorrelationSubscription> correlationSuscription (RuleFlowProcess process) {
+        Map<String, CorrelationSubscription> correlationSuscription = (Map<String, CorrelationSubscription>) process.getMetaData("CorrelationSuscriptions");
+        if(correlationSuscription == null) {
```
[the remark this answers is not in the document]
Comment (elguardian) [reply jbpm-bpmn2/src/main/java/org/jbpm/bpmn2/xml/HandlerUtil.java:72]: done

File: jbpm-bpmn2/src/main/java/org/jbpm/bpmn2/xml/HandlerUtil.java [hunk 14]
```diff
@@ -0,0 +1,78 @@
...
+        Map<String, Collaboration> collaborations = (Map<String, Collaboration>) ((ProcessBuildData) parser.getData()).getMetaData("Collaborations");
+        if(collaborations == null) {
+            collaborations = new HashMap<>();
+            ((ProcessBuildData) parser.getData()).setMetaData("Collaborations", collaborations);
+        }
+        return collaborations;
+    }
+
+    public static Map<String, CorrelationSubscription> correlationSuscription (RuleFlowProcess process) {
+        Map<String, CorrelationSubscription> correlationSuscription = (Map<String, CorrelationSubscription>) process.getMetaData("CorrelationSuscriptions");
```
[the remark this answers is not in the document]
Comment (elguardian) [reply jbpm-bpmn2/src/main/java/org/jbpm/bpmn2/xml/HandlerUtil.java:71]: done

File: jbpm-bpmn2/src/main/java/org/jbpm/bpmn2/xml/HandlerUtil.java [hunk 15]
```diff
@@ -0,0 +1,78 @@
...
+        if(correlationSuscription == null) {
+            correlationSuscription = new HashMap<>();
+            process.setMetaData("CorrelationSuscriptions", correlationSuscription);
```
[the remark this answers is not in the document]
Comment (elguardian) [reply jbpm-bpmn2/src/main/java/org/jbpm/bpmn2/xml/HandlerUtil.java:74]: done

File: jbpm-bpmn2/src/main/java/org/jbpm/bpmn2/xml/StartEventHandler.java [hunk 16]
```diff
@@ -126,7 +128,8 @@ protected void handleNode(final Node node, final Element element, final String u
                 startNode.setMetaData("MessageType", message.getType());
 
 
-                addTriggerWithInMappings(startNode, "Message-" + message.getName());
+
+                addTriggerWithInMappings(startNode, "Message-" + message.getName(), message.getId(), ((RuleFlowProcess) parser.getMetaData().get("CurrentProcessDefintion")).getCorrelationManager());
```
[the remark this answers is not in the document]
Comment (elguardian) [reply jbpm-bpmn2/src/main/java/org/jbpm/bpmn2/xml/StartEventHandler.java:132]: done

File: jbpm-bpmn2/src/main/java/org/jbpm/bpmn2/xml/BoundaryEventHandler.java [hunk 17]
```diff
@@ -445,7 +445,9 @@ protected void handleMessageNode(final Node node, final Element element,
                 eventNode.setMetaData("MessageType", message.getType());
                 List<EventFilter> eventFilters = new ArrayList<EventFilter>();
                 EventTypeFilter eventFilter = new EventTypeFilter();
+                eventFilter.setCorrelationManager(((RuleFlowProcess) parser.getMetaData().get("CurrentProcessDefintion")).getCorrelationManager());
```
[the remark this answers is not in the document]
Comment (elguardian) [reply jbpm-bpmn2/src/main/java/org/jbpm/bpmn2/xml/BoundaryEventHandler.java:448]: done

File: jbpm-bpmn2/src/main/java/org/jbpm/bpmn2/xml/IntermediateCatchEventHandler.java [hunk 18]
```diff
@@ -226,11 +226,14 @@ protected void handleMessageNode(final Node node, final Element element,
                     throw new IllegalArgumentException(
                             "Could not find message " + messageRef);
                 }
+
                 message.addIncomingNode(node);
                 eventNode.setMetaData("MessageType", message.getType());
                 List<EventFilter> eventFilters = new ArrayList<EventFilter>();
                 EventTypeFilter eventFilter = new EventTypeFilter();
+                eventFilter.setCorrelationManager(((RuleFlowProcess) parser.getMetaData().get("CurrentProcessDefintion")).getCorrelationManager());
```
[the remark this answers is not in the document]
Comment (elguardian) [reply jbpm-bpmn2/src/main/java/org/jbpm/bpmn2/xml/IntermediateCatchEventHandler.java:234]: done

File: jbpm-bpmn2/src/main/java/org/jbpm/bpmn2/xml/ProcessHandler.java [hunk 19]
```diff
@@ -175,7 +182,8 @@ public Object start(final String uri, final String localName,
 		
 		// for unique id's of nodes, start with one to avoid returning wrong nodes for dynamic nodes
 		parser.getMetaData().put("idGen", new AtomicInteger(1));
-		
+		parser.getMetaData().put("CurrentProcessDefintion", process);
```
[the remark this answers is not in the document]
Comment (elguardian) [reply jbpm-bpmn2/src/main/java/org/jbpm/bpmn2/xml/ProcessHandler.java:185]: done

File: jbpm-bpmn2/src/main/java/org/jbpm/bpmn2/xml/ProcessHandler.java [hunk 20]
```diff
@@ -202,10 +210,49 @@ public Object end(final String uri, final String localName,
...
 
+    private void postProcessCollaborations(RuleFlowProcess process, ExtensibleXmlParser parser) {
+        // now we wire correlation process subscriptions
+        CorrelationManager correlationManager = process.getCorrelationManager();
+        for(Message message : HandlerUtil.messages(parser).values()) {
+            correlationManager.newMessage(message.getId(), message.getName(), message.getType());
+        }
+
+        // process which keys we are subscribed
+        Collection<CorrelationSubscription> subscriptions = HandlerUtil.correlationSuscription(process).values();
```
[the remark this answers is not in the document]
Comment (elguardian) [reply jbpm-bpmn2/src/main/java/org/jbpm/bpmn2/xml/ProcessHandler.java:226]: removed

File: jbpm-flow/src/main/java/org/jbpm/process/core/correlation/CorrelationManager.java [hunk 21]
```diff
@@ -0,0 +1,144 @@
...
+        return correlationInstance;
+    }
+
+    private Correlation findCorrelationByMessageRef(String messageRef) {
+        for (Correlation correlation : correlations.values()) {
+            if (correlation.hasCorrelationFor(messageRef)) {
+                return correlation;
+            }
+        }
+        throw new IllegalArgumentException("Correlation for message ref " + messageRef + " does not exists");
```
[the remark this answers is not in the document]
Comment (elguardian) [reply jbpm-flow/src/main/java/org/jbpm/process/core/correlation/CorrelationManager.java:131]: done

File: jbpm-flow/src/main/java/org/jbpm/process/core/event/EventTypeFilter.java [hunk 22]
```diff
@@ -46,20 +61,23 @@ public void setType(String type) {
...
-	}
+    public String toString() {
+        return "Event filter: [" + this.type + "]";
+    }
 
     @Override
     public boolean acceptsEvent(String type, Object event, Function<String, Object> resolver) {
-        return this.type != null && isAccepted(type, resolver);
+        if( this.type != null && isAccepted(type, resolver)) {
+            if(correlationManager != null && correlationManager.isSubscribe(messageRef)) {
```
[the remark this answers is not in the document]
Comment (elguardian) [reply jbpm-flow/src/main/java/org/jbpm/process/core/event/EventTypeFilter.java:75]: done

File: jbpm-flow/src/main/java/org/jbpm/process/core/event/MVELMessageExpressionEvaluator.java [hunk 23]
```diff
@@ -0,0 +1,53 @@
...
+    @Override
+    public Object eval(Object event) {
+        return MVEL.eval(expression, event);
+    }
+
+    @Override
+    public Object eval(Function<String, Object> resolver) {
+        ImmutableDefaultFactory factory = new ImmutableDefaultFactory() {
+            private static final long serialVersionUID = 1L;
+            public boolean isResolveable(String name) {
```
[the remark this answers is not in the document]
Comment (elguardian) [reply jbpm-flow/src/main/java/org/jbpm/process/core/event/MVELMessageExpressionEvaluator.java:43]: added

File: jbpm-flow/src/main/java/org/jbpm/process/core/event/MVELMessageExpressionEvaluator.java [hunk 24]
```diff
@@ -0,0 +1,53 @@
...
+                return resolver.apply(name) != null;
+            };
+            public VariableResolver getVariableResolver(String name) {
```
[the remark this answers is not in the document]
Comment (elguardian) [reply jbpm-flow/src/main/java/org/jbpm/process/core/event/MVELMessageExpressionEvaluator.java:46]: added

File: jbpm-test-coverage/src/test/java/org/jbpm/test/functional/collaboration/CollaborationTest.java [hunk 25]
```diff
@@ -0,0 +1,79 @@
...
+    @Test
+    public void testStartMessageCollaboration () {
+        addWorkItemHandler("Human Task", getTestWorkItemHandler());
+        createRuntimeManager("org/jbpm/test/functional/collaboration/Collaboration-StartMessage.bpmn2");
+        RuntimeEngine runtimeEngine = getRuntimeEngine();
+        KieSession ksession = runtimeEngine.getKieSession();
+        ksession.signalEvent("Message-collaboration", new Message("1", "example"));
+        assertEquals(1, getLogService().findProcessInstances().size());
+    }
+
```
[the remark this answers is not in the document]
Comment (elguardian) [reply jbpm-test-coverage/src/test/java/org/jbpm/test/functional/collaboration/CollaborationTest.java:56]: added

File: jbpm-flow/src/main/java/org/jbpm/process/core/correlation/CorrelationManager.java (hunk 21 above)
[the remark this answers is not in the document]
Comment (gmunozfe) [reply jbpm-flow/src/main/java/org/jbpm/process/core/correlation/CorrelationManager.java:131]: I think you didn't commit, seeing still the typo

**Review by gmunozfe — COMMENTED** [review]
Just missing one class to be serializable and a little typo in an exception message

File: jbpm-bpmn2/src/main/java/org/jbpm/bpmn2/core/Collaboration.java (hunk 5 above)
[the remark this answers is not in the document]
Comment (elguardian) [reply jbpm-bpmn2/src/main/java/org/jbpm/bpmn2/core/Collaboration.java:22]: this case is not needed. but ok (it belongs to the parser)

**Comment by elguardian** [comment]
Jenkins retest this please.
